src/cgroups/memory.rs

```diff
@@ -0,0 +1,336 @@
+        OpenOptions::new()
+            .create(false)
+            .write(true)
+            .truncate(true)
+            .open(path)?
+            .write_all(val.to_string().as_bytes())?;
+        Ok(())
+    }
+
+    fn set_u64(val: u64, path: &Path) -> std::io::Result<()> {
+        OpenOptions::new()
+            .create(false)
+            .write(true)
+            .truncate(true)
+            .open(path)?
+            .write_all(val.to_string().as_bytes())?;
+        Ok(())
+    }
+
+    fn set_memory(val: i64, cgroup_root: &Path) -> Result<()> {
+        let path = cgroup_root.join(CGROUP_MEMORY_LIMIT);
+
+        match Self::set_i64(val, &path) {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                // we need to look into the raw OS error for an EBUSY status
+                if let Some(code) = e.raw_os_error() {
+                    // the nix crate has a handy enum for these, lets use that
+                    let errno = Errno::from_i32(code);
+                    // if the error is EBUSY
+                    if let Errno::EBUSY = errno {
+                        let usage = Self::get_memory_usage(cgroup_root)?;
+                        let max_usage = Self::get_memory_max_usage(cgroup_root)?;
+                        Err(anyhow!(
+                            "unable to set memory limit to {} (current usage: {}, peak usage: {})",
+                            val,
+                            usage,
+                            max_usage,
+                        ))
+                    } else {
+                        Err(anyhow!(e))
+                    }
+                } else {
+                    Err(anyhow!(e))
+                }
+            }
+        }
+    }
+
+    fn set_swap(val: i64, cgroup_root: &Path) -> Result<()> {
+        if val == 0 {
+            return Ok(());
+        }
+
+        let path = cgroup_root.join(CGROUP_MEMORY_SWAP_LIMIT);
+
+        Self::set_i64(val, &path)?;
+
+        Ok(())
+    }
+
+    fn set_memory_and_swap(resource: &LinuxMemory, cgroup_root: &Path) -> Result<()> {
+        let limit = resource.limit.unwrap_or(-1);
+        let mut swap = resource.swap.unwrap_or(0);
+
+        if limit == -1 && swap == 0 {
+            if cgroup_root.join(CGROUP_MEMORY_SWAP_LIMIT).exists() {
+                swap = -1;
+            }
+        }
+
+        // According to runc we need to change the write sequence of
+        // limit and swap so it won't fail, because the new and old
+        // values don't fit the kernel's validation
+        // see:
+        // https://github.com/opencontainers/runc/blob/master/libcontainer/cgroups/fs/memory.go#L89
+        if limit != 0 && swap != 0 {
+            let current_limit = Self::get_memory_limit(cgroup_root)?;
+
+            if swap == -1 || current_limit < swap {
+                Self::set_swap(swap, cgroup_root)?;
+                Self::set_memory(limit, cgroup_root)?;
+                return Ok(());
+            }
```

Comment:
I think it would be easier to use Option to make the code easier.
Also, `swap `can turn off `mut` by determining the value at the time of allocation.